include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+   * scalar FiniteElement @projection_fe, @p mass_quadrature and @p data_quadrature
+   * Quadrature rules.
+   * First, the data located at @p data_quadrature of each cell is L2-projected
+   * to the continuous space defined by a single FiniteElement @projection_fe .
+   * In  doing so the mass matrix of this element is required, which will be calculated
+   * with the @p mass_quadrature rule . Should the cell now belong to another processor,
+   * the data is then sent to this processor. The class makes use of a feature of p4est
+   * (and parallel::distributed::Triangulation) that allows one to attach information
+   * to cells during mesh refinement and rebalancing.
+   * On receiving information on the target cell, the data is projected back to the
+   * quadrature points. In the case that local refinement is performed, we have to first
+   * project local DoF values of the parent element to each child.
+   *
+   * This class is templated by @p DataType type, however the user's @p DataType class
+   * has to be derived from the TransferableQuadraturePointData class. In practice
+   * that amounts to implementing the following three functions shown below for a
+   * quadrature point data with 2 scalars :
+   * <code>
+   * class MyQData : public TransferableQuadraturePointData
+   * {
+   * public:
+   *   unsigned int number_of_values() const
+   *   {
+   *      return 2;
+   *   }
+   *
+   *   // a function to pack scalars into a vector
+   *   void pack_values(std::vector<double> &values) const
+   *   {
+   *     Assert (scalars.size()==2, ExcInternalError());
+   *     scalars[0] = value1;
+   *     scalars[1] = value2;
+   *   }
+   *
+   *   void unpack_values(const std::vector<double> &values)
+   *   {
+   *     Assert (scalars.size() ==2, ExcInternalError());
+   *     value1 = scalars[0];
+   *     value2 = scalars[1];
+   *   }
+   * };
+   * </code>
+   * Note that the order of packing and unpacking has to be the same.
+   *
+   * This class can then be use with CellDataStorage in the following way:
+   * <code>
+   * CellDataStorage<typename Triangulation<dim,dim>::cell_iterator,MyQData> data_storage;
+   * distributed::ContinuousQuadratureDataTransfer<dim,MyQData> data_transfer(FE_Q<dim>(2),QGauss<dim>(3),QGauss<dim>(4));
+   * //...populate data for all active cells in data_storage
+   * //...mark cells for refinement...
+   * data_transfer.prepare_for_coarsening_and_refinement(triangulation,&data_storage);
+   * triangulation.execute_coarsening_and_refinement();
+   * //...initialize quadrature point data on new cells by calling CellDataStorage::reinit()
+   * data_transfer.interpolate();
+   * </code>
+   * This approach can be extended to quadrature point data with Tensors of
+   * arbitrary order, although with a little bit more work in packing and
+   * unpacking of data inside MyQData class.
+   *
+   * @note Currently coarsening is not supported.
```

Comment:
On 05/06/2016 07:46 AM, Denis Davydov wrote:
> Will the data at children always be available (say of some children were
> on different processors and now they are coarsened into a single cell on
> a single processor)?

I suspect that if you know that the children of a parent cell are going 
to be coarsened away, and at least one of the children is locally owned, 
then the others are definitely ghost cells. So the data should be available.